src/components/Search/SearchSelectionListWithHeader.tsx

```diff
@@ -0,0 +1,124 @@
+            if (!item.keyForList) {
+                return;
+            }
+
+            setSelectedItems((prev) => {
+                if (prev[item.keyForList]?.isSelected) {
+                    const {[item.keyForList]: omittedTransaction, ...transactions} = prev;
+                    return transactions;
+                }
+                return {...prev, [item.keyForList]: {isSelected: true, canDelete: item.canDelete, action: item.action}};
+            });
+
+            return;
+        }
+
+        if (item.transactions.every((transaction) => selectedItems[transaction.keyForList]?.isSelected)) {
+            const reducedSelectedItems: SelectedTransactions = {...selectedItems};
+
+            item.transactions.forEach((transaction) => {
+                delete reducedSelectedItems[transaction.keyForList];
+            });
+
+            setSelectedItems(reducedSelectedItems);
+            return;
+        }
+
+        setSelectedItems({
+            ...selectedItems,
+            ...Object.fromEntries(item.transactions.map((transaction) => [transaction.keyForList, {isSelected: true, canDelete: transaction.canDelete, action: transaction.action}])),
+        });
+    };
+
+    const toggleAllTransactions = () => {
+        const areReportItems = searchType === CONST.SEARCH.DATA_TYPES.REPORT;
+        const flattenedItems = areReportItems ? (data as ReportListItemType[]).flatMap((item) => item.transactions) : data;
+        const isAllSelected = flattenedItems.length === Object.keys(selectedItems).length;
+
+        if (isAllSelected) {
+            clearSelectedItems();
+            return;
+        }
+
+        if (areReportItems) {
+            setSelectedItems(
+                Object.fromEntries(
+                    (data as ReportListItemType[]).flatMap((item) =>
+                        item.transactions.map((transaction: TransactionListItemType) => [
+                            transaction.keyForList,
+                            {isSelected: true, canDelete: transaction.canDelete, action: transaction.action},
+                        ]),
+                    ),
+                ),
+            );
+
+            return;
+        }
+
+        setSelectedItems(Object.fromEntries((data as TransactionListItemType[]).map((item) => [item.keyForList, {isSelected: true, canDelete: item.canDelete, action: item.action}])));
+    };
+
+    const mapToSelectedTransactionItem = (item: TransactionListItemType) => ({...item, isSelected: !!selectedItems[item.keyForList]?.isSelected});
+
+    const sortedSelectedData = data.map((item) =>
+        SearchUtils.isTransactionListItemType(item)
+            ? mapToSelectedTransactionItem(item)
+            : {
+                  ...item,
+                  transactions: item.transactions?.map(mapToSelectedTransactionItem),
+                  isSelected: item.transactions.every((transaction) => !!selectedItems[transaction.keyForList]?.isSelected),
+              },
```

Comment:
I would rather have a function that accepts base `item` and handles all the logic inside it.
Because right now we have an `if` expression and 1 branch is handled via dedicated function but the other branch is handled in place by creating a new item.

```
const sortedSelectedData = data.map(mapToSelectedItem) // <-- do all the work inside `mapToSelectedItem`
```